SQL connection lag issue on 1.18.1

Appears to be happening in the context of when [this bug](https://bugs.mojang.com/browse/MC-227945) is causing the server to lag badly

```
13.01 19:41:20 [Server] INFO Geph issued server command: /oxygen
13.01 19:41:20 [Server] WARN com.mysql.cj.jdbc.exceptions.CommunicationsException: The last packet successfully received from the server was 90,439,088 milliseconds ago. The last packet sent successfully to the server was 90,439,088 milliseconds ago. is longer than the server configured value of 'wait_timeout'. You should consider either expiring and/or testing connection validity before use in your application, increasing the server configured values for client timeouts, or using the Connector/J connection property 'autoReconnect=true' to avoid this problem.
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.exceptions.SQLError.createCommunicationsException(SQLError.java:174)
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:64)
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:953)
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1098)
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1046)
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1371)
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdate(ClientPreparedStatement.java:1031)
13.01 19:41:20 [Server] WARN at WHIMC-ScienceTools-2.4.5.jar//edu.whimc.sciencetools.utils.sql.Queryer.lambda$storeNewMeasurement$4(Queryer.java:115)
13.01 19:41:20 [Server] WARN at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftTask.run(CraftTask.java:101)
13.01 19:41:20 [Server] WARN at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57)
13.01 19:41:20 [Server] WARN at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22)
13.01 19:41:20 [Server] WARN at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
13.01 19:41:20 [Server] WARN at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
13.01 19:41:20 [Server] WARN at java.base/java.lang.Thread.run(Thread.java:833)
13.01 19:41:20 [Server] WARN Caused by: com.mysql.cj.exceptions.CJCommunicationsException: The last packet successfully received from the server was 90,439,088 milliseconds ago. The last packet sent successfully to the server was 90,439,088 milliseconds ago. is longer than the server configured value of 'wait_timeout'. You should consider either expiring and/or testing connection validity before use in your application, increasing the server configured values for client timeouts, or using the Connector/J connection property 'autoReconnect=true' to avoid this problem.
13.01 19:41:20 [Server] WARN at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
13.01 19:41:20 [Server] WARN at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
13.01 19:41:20 [Server] WARN at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
13.01 19:41:20 [Server] WARN at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
13.01 19:41:20 [Server] WARN at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
13.01 19:41:20 [Server] WARN at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61)
13.01 19:41:20 [Server] WARN at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105)
13.01 19:41:20 [Server] WARN at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:151)
13.01 19:41:20 [Server] WARN at com.mysql.cj.exceptions.ExceptionFactory.createCommunicationsException(ExceptionFactory.java:167)
13.01 19:41:20 [Server] WARN at com.mysql.cj.protocol.a.NativeProtocol.send(NativeProtocol.java:567)
13.01 19:41:20 [Server] WARN at com.mysql.cj.protocol.a.NativeProtocol.sendCommand(NativeProtocol.java:622)
13.01 19:41:20 [Server] WARN at com.mysql.cj.protocol.a.NativeProtocol.sendQueryPacket(NativeProtocol.java:987)
13.01 19:41:20 [Server] WARN at com.mysql.cj.NativeSession.execSQL(NativeSession.java:666)
13.01 19:41:20 [Server] WARN at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:930)
13.01 19:41:20 [Server] WARN ... 11 more
13.01 19:41:20 [Server] WARN Caused by: java.net.SocketException: Broken pipe
13.01 19:41:20 [Server] WARN at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:420)
13.01 19:41:20 [Server] WARN at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:440)
13.01 19:41:20 [Server] WARN at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:826)
13.01 19:41:20 [Server] WARN at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1035)
13.01 19:41:20 [Server] WARN at java.base/java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:81)
13.01 19:41:20 [Server] WARN at java.base/java.io.BufferedOutputStream.flush(BufferedOutputStream.java:142)
13.01 19:41:20 [Server] WARN at com.mysql.cj.protocol.a.SimplePacketSender.send(SimplePacketSender.java:55)
13.01 19:41:20 [Server] WARN at com.mysql.cj.protocol.a.TimeTrackingPacketSender.send(TimeTrackingPacketSender.java:50)
13.01 19:41:20 [Server] WARN at com.mysql.cj.protocol.a.NativeProtocol.send(NativeProtocol.java:558)
13.01 19:41:20 [Server] WARN ... 15 more
```

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
     <modelVersion>4.0.0</modelVersion>
     <groupId>edu.whimc</groupId>
     <artifactId>WHIMC-ScienceTools</artifactId>
-    <version>2.4.5</version>
+    <version>2.4.6</version>
     <name>WHIMC Science Tools</name>
     <description>Simulate values for scientific tools</description>
     <repositories>

--- a/src/main/java/edu/whimc/sciencetools/utils/sql/MySQLConnection.java
+++ b/src/main/java/edu/whimc/sciencetools/utils/sql/MySQLConnection.java
@@ -1,27 +1,21 @@
 package edu.whimc.sciencetools.utils.sql;
 
+import com.mysql.cj.jdbc.MysqlConnectionPoolDataSource;
+import com.mysql.cj.jdbc.MysqlDataSource;
 import edu.whimc.sciencetools.ScienceTools;
 import edu.whimc.sciencetools.utils.sql.migration.SchemaManager;
 import java.sql.Connection;
-import java.sql.DriverManager;
 import java.sql.SQLException;
+import org.bukkit.configuration.ConfigurationSection;
 
 /**
  * Handles creating connections to the configured MySQL database.
  */
 public class MySQLConnection {
 
-    private static final String URL_TEMPLATE = "jdbc:mysql://%s:%s/%s";
-
-    private final String host;
-    private final String database;
-    private final String username;
-    private final String password;
-    private final String url;
-    private final int port;
     private final ScienceTools plugin;
 
-    private Connection connection;
+    private MysqlDataSource dataSource;
 
     /**
      * Create an instance of {@link MySQLConnection}.
@@ -29,15 +23,15 @@ public class MySQLConnection {
      * @param plugin The plugin instance.
      */
     public MySQLConnection(ScienceTools plugin) {
-        this.host = plugin.getConfig().getString("mysql.host", "localhost");
-        this.port = plugin.getConfig().getInt("mysql.port", 3306);
-        this.database = plugin.getConfig().getString("mysql.database", "minecraft");
-        this.username = plugin.getConfig().getString("mysql.username", "user");
-        this.password = plugin.getConfig().getString("mysql.password", "pass");
-
-        this.url = String.format(URL_TEMPLATE, this.host, this.port, this.database);
-
         this.plugin = plugin;
+        this.dataSource = new MysqlConnectionPoolDataSource();
+
+        ConfigurationSection config = plugin.getConfig();
+        this.dataSource.setServerName(config.getString("mysql.host", "localhost"));
+        this.dataSource.setPortNumber(config.getInt("mysql.port", 3306));
+        this.dataSource.setDatabaseName(config.getString("mysql.database", "minecraft"));
+        this.dataSource.setUser(config.getString("mysql.username", "user"));
+        this.dataSource.setPassword(config.getString("mysql.password", "pass"));
     }
 
     /**
@@ -46,29 +40,28 @@ public class MySQLConnection {
      * @return Whether the database was successfully initialized.
      */
     public boolean initialize() {
-        if (getConnection() == null) {
+        try (Connection connection = getConnection()) {
+            if (connection == null) {
+                return false;
+            }
+
+            SchemaManager manager = new SchemaManager(this.plugin, connection);
+            return manager.migrate();
+        } catch (SQLException unused) {
             return false;
         }
-
-        SchemaManager manager = new SchemaManager(this.plugin, this.connection);
-        return manager.migrate();
     }
 
     /**
      * A connection to the configured database or null if unsuccessful.
      */
-    public Connection getConnection() {
-        try {
-            if (this.connection != null && !this.connection.isClosed()) {
-                return this.connection;
-            }
-
-            this.connection = DriverManager.getConnection(this.url, this.username, this.password);
-        } catch (SQLException unused) {
+    public Connection getConnection() throws SQLException {
+        Connection connection = this.dataSource.getConnection();
+        if (!connection.isValid(1)) {
             return null;
         }
 
-        return this.connection;
+        return connection;
     }
 
 }

